core/serialize/serialize.cc

```diff
@@ -911,25 +906,19 @@ CachedFile Serializer::loadFile(const core::GlobalState &gs, core::FileRef fref,
     return CachedFile{move(file), move(tree)};
 }
 
-void SerializerImpl::pickleTree(Pickler &p, ast::TreePtr &t) {
-    pickle(p, t);
-}
-
-void SerializerImpl::pickleAstHeader(Pickler &p, u1 tag, core::LocOffsets loc) {
-    p.putU1(tag);
-    pickle(p, loc);
-}
-
 void SerializerImpl::pickle(Pickler &p, const ast::TreePtr &what) {
     if (what == nullptr) {
-        p.putU1(1);
+        p.putU4(0);
         return;
     }
 
-    typecase(
-        what.get(),
-        [&](ast::Send *s) {
-            pickleAstHeader(p, 2, s->loc);
+    auto tag = what.tag();
+    p.putU4(static_cast<u4>(tag));
+
+    switch (tag) {
+        case ast::Tag::Send: {
+            auto *s = ast::cast_tree<ast::Send>(what);
```

Comment:
You're paying the cost of two tag checks here. A nonnull cast is 'free'.

```suggestion
            auto &s = ast::cast_tree_nonnull<ast::Send>(what);
```

Reply:
Yes, but this was the least invasive refactoring. I guess given how large the diff already is, it's probably not a big deal to change a bunch of `->` to `.`.